ompi/mca/coll/han/coll_han_scatterv.c

```diff
@@ -0,0 +1,388 @@
+         * temp buf without gaps */
+        for (up_peer = 0; up_peer < up_size; ++up_peer) {
+            if (root_up_rank == up_peer) {
+                continue;
+            }
+            if (!need_bounce_buf && 0 < up_scounts[up_peer]
+                && up_scounts[up_peer] != up_peer_ub[up_peer] - up_displs[up_peer]) {
+                need_bounce_buf = 1;
+                break;
+            }
+        }
+
+        if (need_bounce_buf) {
+            bounce_buf = malloc(sdsize * total_up_scounts);
+            if (!bounce_buf) {
+                err = OMPI_ERR_OUT_OF_RESOURCE;
+                goto root_out;
+            }
+
+            /* Calculate displacements for the inter-node scatterv */
+            for (up_peer = 0; up_peer < up_size; ++up_peer) {
+                up_displs[up_peer] = 0 < up_peer ? up_displs[up_peer - 1] + up_scounts[up_peer - 1]
+                                                 : 0;
+            }
+
+            /* Use a temp buffer to reorder the send buffer if needed */
+            ptrdiff_t offset = 0;
+
+            for (int i = 0; i < w_size; ++i) {
+                up_peer = topo[2 * i];
+                if (root_up_rank == up_peer) {
+                    continue;
+                }
+
+                w_peer = topo[2 * i + 1];
+
+                ompi_datatype_copy_content_same_ddt(sdtype, (size_t) scounts[w_peer],
+                                                    bounce_buf + offset,
+                                                    (char *) sbuf
+                                                        + (size_t) displs[w_peer] * sdsize);
+                offset += sdsize * (size_t) scounts[w_peer];
+            }
+
+            reorder_sbuf = bounce_buf;
+        }
+
+        /* Up Iscatterv */
+        up_comm->c_coll->coll_iscatterv((const char *) reorder_sbuf, up_scounts, up_displs, sdtype,
+                                        rbuf, rcount, rdtype, root_up_rank, up_comm, &iscatterv_req,
+                                        up_comm->c_coll->coll_scatterv_module);
+
+        /* Low Scatterv */
+        low_comm->c_coll->coll_scatterv(sbuf, low_scounts, low_displs, sdtype, rbuf, rcount, rdtype,
+                                        root_low_rank, low_comm,
+                                        low_comm->c_coll->coll_scatterv_module);
+
+        ompi_request_wait(&iscatterv_req, MPI_STATUS_IGNORE);
+
+    root_out:
+        if (low_displs) {
+            free(low_displs);
+        }
+        if (low_scounts) {
+            free(low_scounts);
+        }
+        if (up_displs) {
+            free(up_displs);
+        }
+        if (up_scounts) {
+            free(up_scounts);
+        }
+        if (up_peer_lb) {
+            free(up_peer_lb);
+        }
+        if (up_peer_ub) {
+            free(up_peer_ub);
+        }
+        if (bounce_buf) {
+            free(bounce_buf);
+        }
+        if (iscatterv_req) {
+            ompi_request_free(&iscatterv_req);
+        }
```

Comment:
Not needed because wait will free the request.

Reply:
Thanks! I didn't realize that. Fixed.